eth_account/_utils/encode_typed_data/encoding_and_hashing.py

```diff
@@ -0,0 +1,209 @@
+from typing import (
+    Any,
+    Dict,
+    List,
+    Tuple,
+    Union,
+)
+
+from eth_abi import (
+    encode,
+)
+from eth_utils import (
+    is_hexstr,
+    keccak,
+    to_bytes,
+    to_int,
+)
+
+from eth_account._utils.encode_typed_data.helpers import (
+    EIP712_SOLIDITY_TYPES,
+    coerce_bool,
+    is_array_type,
+    parse_core_array_type,
+    parse_parent_array_type,
+)
+
+
+def get_primary_type(types: Dict[str, List[Dict[str, str]]]) -> str:
+    custom_types = set(types.keys())
+    custom_types_that_are_deps = set()
+
+    for type_ in custom_types:
+        type_fields = types[type_]
+        for field in type_fields:
+            parsed_type = parse_core_array_type(field["type"])
+            if parsed_type in custom_types and parsed_type != type_:
+                custom_types_that_are_deps.add(parsed_type)
+
+    primary_type = list(custom_types.difference(custom_types_that_are_deps))
+    if len(primary_type) == 1:
+        return primary_type[0]
+    else:
+        raise ValueError("Unable to determine primary type")
+
+
+def encode_field(
+    types: Dict[str, List[Dict[str, str]]],
+    name: str,
+    type_: str,
+    value: Any,
+) -> Tuple[str, Union[int, bytes]]:
+    if type_ in types.keys():
+        # type is a custom type
+        if value is None:
+            return ("bytes32", b"\x00" * 32)
+        else:
+            return ("bytes32", keccak(encode_data(type_, types, value)))
+
+    elif type_ in ["string", "bytes"] and value is None:
+        return ("bytes32", b"")
+
+    # None is allowed only for custom and dynamic types
+    elif value is None:
+        raise ValueError(f"Missing value for field `{name}` of type `{type_}`")
+
+    elif is_array_type(type_):
+        type_ = parse_parent_array_type(type_)
+        type_value_pairs = [encode_field(types, name, type_, item) for item in value]
+        data_types, data_hashes = zip(*type_value_pairs)
+        return ("bytes32", keccak(encode(data_types, data_hashes)))
+
+    elif type_ == "bool":
+        return (type_, coerce_bool(value))
+
+    elif type_ == "bytes":
```

Comment:
Summarizing what we talked about and trying things on my end, comparing with `eth-sig-util` tests, this should include the following logic in one way or another:

```py
    if is_array_type(type_):
        ...
    elif type_ == "bool":
        ...
    elif type_.startswith("bytes") and not is_array_type(type_):
        if not isinstance(value, bytes):
            value = (
                to_bytes(hexstr=value)
                if is_0x_prefixed_hexstr(value)
                else to_bytes(text=value)
                if isinstance(value, str)
                else to_bytes(value)
            )

        return (
            # keccak hash if dynamic `bytes` type
            ("bytes32", keccak(value))
            if type_ == "bytes"
            # if fixed bytesXX type, do not hash
            else ("bytes32", value)
        )
```

Where the helper method `is_0x_prefixed_hexstr()` checks for the `0x` at the beginning. Otherwise, strings are counted as a hexstring if they pass the regex [here](https://github.com/ethereum/eth-utils/blob/e1a395207f26962e49147fc80ccff23f150b2d62/eth_utils/hexadecimal.py#L11-L11), which doesn't require the `0x`. This means a string that is meant to be a string representation of the number "10" is treated as a hex string instead. "0x10" is very different from "10", which is why those tests were failing.

On my end, I included the `is_0x_prefixed_hexstr()` method in your `encode_typed_data/helpers.py` file and it just looks like this:

```py
def is_0x_prefixed_hexstr(value: Any):
    return is_hexstr(value) and value.startswith("0x")
```

Reply:
good stuff, added